File: Nikhil/java/core/Agent.java
package core;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.openai.client.OpenAIClient;
import com.openai.client.okhttp.OpenAIOkHttpClient;
import com.openai.core.JsonMissing;
import com.openai.models.FunctionDefinition;
import com.openai.models.FunctionParameters;
import com.openai.models.ResponseFormatJsonSchema;
import com.openai.models.ResponseFormatJsonSchema.JsonSchema.Schema;
import com.openai.models.chat.completions.ChatCompletionAssistantMessageParam;
import com.openai.models.chat.completions.ChatCompletionAssistantMessageParam.Builder;
import com.openai.models.chat.completions.ChatCompletionCreateParams;
import com.openai.models.chat.completions.ChatCompletionCreateParams.ResponseFormat;
import com.openai.models.chat.completions.ChatCompletionDeveloperMessageParam;
import com.openai.models.chat.completions.ChatCompletionMessage;
import com.openai.models.chat.completions.ChatCompletionMessageParam;
import com.openai.models.chat.completions.ChatCompletionMessageToolCall;
import com.openai.models.chat.completions.ChatCompletionMessageToolCall.Function;
import com.openai.models.chat.completions.ChatCompletionTool;
import com.openai.models.chat.completions.ChatCompletionToolMessageParam;
import com.openai.models.chat.completions.ChatCompletionUserMessageParam;

import lombok.Getter;
import lombok.NonNull;
import lombok.Setter;

/**
 * This class implements an Agent that uses OpenAI models via the Chat
 * Completions API. This class is implemented using the OpenAI Java SDK.
 * 
 * @author Massimiliano "Maxi" Zattera
 */
public class Agent {

	private final static Logger LOG = LoggerFactory.getLogger(Agent.class);

	public static final String DEFAULT_MODEL = "gpt-4.1";

	protected Agent() {
		this("OpenAIChatCompletionService", "Test agent", new ArrayList<>());
	}

	protected Agent(@NonNull String id, @NonNull String description, @NonNull List<? extends Tool> tools) {
		this.client = OpenAIOkHttpClient.fromEnv();
		this.id = id;
		this.description = description;
		for (Tool t : tools) {
			t.init(this);
			toolMap.put(t.getId(), t);
		}
		this.model = DEFAULT_MODEL;
		this.temperature = 0.0d;
	}

	/**
	 * This class uses an OpenAIClient to call the OpenAI API. In python you do not
	 * need a client; you can omit this field and use openai intead.
	 * 
	 * E.g. instead of:
	 * 
	 * client.chat().completions().create(req);
	 * 
	 * In Python use:
	 * 
	 * import openai
	 * 
	 * openai.ChatCompletion.create(**req)
	 */
	protected final OpenAIClient client;

	/**
	 * Unique agent ID.
	 */
	@Getter
	private final String id;

	/**
	 * A verbose description of the agent, describing its capabilities.
	 */
	@Getter
	@Setter
	private String description;

	/**
	 * The name of the LLM model used by this agent (e.g. "gpt-4").
	 */
	@Getter
	@Setter
	private String model;

	/**
	 * Personality (system prompt) for the agent.
	 */
	@Getter
	@Setter
	private String personality;

	/**
	 * Messages exchanged so far in current conversation with the agent.
	 */
	@Getter
	private final List<ChatMessage> history = new ArrayList<>();

	/**
	 * The temperature of the sampling operation for the underlying LLM.
	 */
	@Getter
	@Setter
	private Double temperature = 0d;

	/**
	 * Starts a new chat, clearing current conversation.
	 */
	public void clearConversation() {
		history.clear();
	}

	/**
	 * Maximum number of messages to keep in chat history.
	 */
	@Getter
	@Setter
	private int maxHistoryLength = Integer.MAX_VALUE;

	/**
	 * Maximum number of history messages to send to LLM during a conversation.
	 * 
	 * Notice this does NOT limit length of conversation history (see
	 * {@link #getMaxHistoryLength}).
	 */
	@Getter
	@Setter
	private int maxConversationSteps = Integer.MAX_VALUE;

	/**
	 * Output format for the model, as JSON schema.
	 */
	@Getter
	private String responseFormat;

	/**
	 * This method allows to specify an output format for the model. This is used to
	 * create structured outputs with model supporting it so that, for example, the
	 * model returns its responses in a pre-defined JSON format.
	 * 
	 * @param schema A class which schema will be used to define the output format.
	 */
	public void setResponseFormat(Class<?> schema) {
		responseFormat = JsonSchema.getJsonSchema(schema);
	}

	/**
	 * Available and already initialized tools that agent can use. Maps each tool ID
	 * into corresponding tool instance.
	 */
	protected Map<String, Tool> toolMap = new HashMap<>();

	/**
	 * Continues current chat, with the provided message.
	 * 
	 * The exchange is added to the conversation history.
	 */
	public ChatCompletion chat(String msg) {
		return chat(new ChatMessage(ChatMessage.Author.USER, msg));

	}

	/**
	 * Continues current chat, with the provided message.
	 * 
	 * The exchange is added to the conversation history.
	 */
	public ChatCompletion chat(ChatMessage msg) {
		return chat(List.of(msg));
	}

	/**
	 * Continues current chat, with the provided messages.
	 * 
	 * The exchange is added to the conversation history.
	 */
	public ChatCompletion chat(List<ChatMessage> msg) {

		// Add messages to conversation and trims it; this also adds personality
		List<ChatMessage> conversation = new ArrayList<>(getHistory());
		conversation.addAll(msg);
		trimConversation(conversation);

		// Create response
		ChatCompletion result = chatCompletion(conversation);

		// Add messages and response to history
		getHistory().addAll(msg);
		getHistory().add(result.getMessage());

		// Make sure history is of desired length
		if (getHistory().size() > getMaxHistoryLength()) {
			getHistory().subList(0, getHistory().size() - getMaxHistoryLength()).clear();
		}

		return result;
	}

	/**
	 * Completes text outside a conversation (executes given prompt ignoring and
	 * without affecting conversation history).
	 * 
	 * Notice this does not consider or affects chat history but bot personality is
	 * used, if provided.
	 */
	public ChatCompletion complete(String prompt) {
		return complete(new ChatMessage(ChatMessage.Author.USER, prompt));
	}

	/**
	 * Completes text outside a conversation (executes given prompt ignoring and
	 * without affecting conversation history).
	 * 
	 * Notice this does not consider or affects chat history but bot personality is
	 * used, if provided.
	 */
	public ChatCompletion complete(ChatMessage prompt) {

		List<ChatMessage> conversation = List.of(prompt);
		trimConversation(conversation);

		return chatCompletion(conversation);
	}

	/**
	 * Trims given list of messages (typically a conversation history), so it fits
	 * the limits set in this instance (that is, maximum conversation steps).
	 * 
	 * Notice the personality (system prompt) is always and automatically added on
	 * top of the trimmed list (if set).
	 * 
	 * @throws IllegalArgumentException if no message can be added because of
	 *                                  context size limitations.
	 */
	private void trimConversation(List<ChatMessage> messages) {

		// Remove tool call results left on top without corresponding calls, or this
		// will cause HTTP 400 error for tools
		int firstNonToolIndex = 0;
		for (ChatMessage m : messages) {
			if (m.hasToolCallResults()) {
				firstNonToolIndex++;
			} else {
				break;
			}
		}
		if (firstNonToolIndex > 0) {
			messages.subList(0, firstNonToolIndex).clear();
		}

		// Trims down the list of messages accordingly to given limits.
		if (messages.size() > getMaxConversationSteps())
			messages.subList(0, messages.size() - getMaxConversationSteps()).clear();

		if (messages.size() == 0)
			throw new IllegalArgumentException("No messages left in conversation");

		if (getPersonality() != null)
			// must add a system message on top with personality
			messages.add(0, new ChatMessage(ChatMessage.Author.DEVELOPER, getPersonality()));
	}

	/**
	 * This method sends given list of messages to the LLM and creates corresponding
	 * chat response, which is returned.
	 */
	@SuppressWarnings("unchecked")
	private ChatCompletion chatCompletion(List<ChatMessage> messages) {

		// The method starts by building required ChatCompletionCreateParams instance to
		// call the OpenAI chat API.
		// When translating this into Python you should be a request req that is then
		// sent to openai.ChatCompletion.create(**req)

		// Translate messages into format suitable to be used with OpenAI API
		// Here each ChatMessage is translated into the corresponding
		// ChatCompletionMessageParam.
		List<ChatCompletionMessageParam> openAiMessages = new ArrayList<>();
		for (ChatMessage m : messages)
			openAiMessages.addAll(fromChatMessage(m));

		// ThHere we use a Builder to build ChatCompletionCreateParams step by step
		// Notice that here we set the model, the messages and the temperature
		// With OpenAI Java SDK, messages must be a list of ChatCompletionMessageParam
		// but in Python you might build them differently as part of req; check what
		// fromChatMessage() does
		ChatCompletionCreateParams.Builder b = ChatCompletionCreateParams.builder() //
				.model(getModel()) //
				.messages(openAiMessages) //
				.temperature(getTemperature());

		// Only if response format was set, provide it; check what
		// createResponseFormat() does
		if (getResponseFormat() == null)
			b.responseFormat(JsonMissing.of());
		else
			b.responseFormat(createResponseFormat());

		// If there is any tool, provide them in req; check what createToolDefinitions()
		// does
		if (toolMap.size() == 0)
			b.tools(JsonMissing.of());
		else
			b.tools(createToolDefinitions());
		ChatCompletionCreateParams req = b.build();

		LOG.info(req.toString());

		// Calls OpenAI Chat Completion API and returns result
		com.openai.models.chat.completions.ChatCompletion resp = client.chat().completions().create(req);
		com.openai.models.chat.completions.ChatCompletion.Choice choice = resp.choices().get(0);
		return new ChatCompletion(fromOpenAiApi(choice.finishReason()), fromOpenAiMessage(choice.message()));
	}

	/**
	 * Translates Java OpenAI SDK finish reason into one we can use.
	 */
	private static ChatCompletion.FinishReason fromOpenAiApi(
			com.openai.models.chat.completions.ChatCompletion.Choice.FinishReason finishReason) {
		switch (finishReason.value()) {
		case STOP:
		case TOOL_CALLS:
		case FUNCTION_CALL:
			return ChatCompletion.FinishReason.COMPLETED;
		case LENGTH:
			return ChatCompletion.FinishReason.TRUNCATED;
		case CONTENT_FILTER:
			return ChatCompletion.FinishReason.INAPPROPRIATE;
		default:
			throw new IllegalArgumentException("Unrecognized finish reason: " + finishReason);
		}
	}

	/**
	 * Turns an ChatCompletionMessageParam returned by API into a ChatMessage. This
	 * is required because we want to return ChatMessage objects and store them as
	 * such in the history.
	 */
	private ChatMessage fromOpenAiMessage(ChatCompletionMessage msg) {

		if (msg.toolCalls().isPresent()) {

			// The model returned a set of tool calls, transparently translate that into a
			// message with a multiple parts each being a ToolCall
			List<ToolCall> calls = new ArrayList<>();
			for (ChatCompletionMessageToolCall call : msg.toolCalls().get()) {
				ToolCall toolCall;
				try {
					toolCall = ToolCall.builder() //
							.id(call.id()) //
							.tool(toolMap.get(call.function().name())) //
							.arguments(call.function().arguments()) //
							.build();
				} catch (JsonProcessingException e) {
					throw new IllegalArgumentException(e);
				}
				calls.add(toolCall);
			}

			// Nothing else to return in this case
			return new ChatMessage(ChatMessage.Author.BOT, calls);
		}

		List<MessagePart> parts = new ArrayList<>();
		if (msg.content().isPresent()) {
			// Normal (text) message was returned, just add it as TextPart
			parts.add(new TextPart(msg.content().get()));
		}
		if (msg.refusal().isPresent()) {
			// OpenAI returned a refusal; translate it into a text message and attach it as
			// TextPart
			parts.add(new TextPart("**The model generated a refusal**\n\n" + msg.refusal().get()));
		}

		// Return a message with all the parts we found
		return new ChatMessage(ChatMessage.Author.BOT, parts);
	}

	/**
	 * This converts a generic ChatMessaege provided by user into an
	 * ChatCompletionMessageParam that is used for the OpenAi API. This is needed
	 * because we want to deal with ChatMessage when talking to the agent.
	 */
	private List<ChatCompletionMessageParam> fromChatMessage(ChatMessage msg) {

		if (msg.hasToolCalls()) {

			// The message contains tool calls; we translate them into a
			// ChatCompletionAssistantMessageParam which will contain all these calls
			Builder b = ChatCompletionAssistantMessageParam.builder();
			for (ToolCall c : msg.getToolCalls()) {
				try {
					b.addToolCall(ChatCompletionMessageToolCall.builder() //
							.id(c.getId()).function(Function.builder() //
									.name(c.getTool().getId()) //
									.arguments(JsonSchema.serialize(c.getArguments())) //
									.build()) //
							.build());
				} catch (JsonProcessingException e) {
					// Rethrows as runtime exception; in Python you do not need to know since you do
					// not need to declare exceptions in methods
					throw new RuntimeException(e);
				}
			}

			// There is nothing else to return since in this case the message will only have
			// calls.
			return List.of(ChatCompletionMessageParam.ofAssistant(b.build())); //
		}

		if (msg.hasToolCallResults()) {

			// The message contains some call results that we translate into corresponding
			// ChatCompletionMessageParam
			List<ChatCompletionMessageParam> result = new ArrayList<>();
			List<ToolCallResult> results = msg.getToolCallResults();
			for (ToolCallResult r : results) {
				result.add(ChatCompletionMessageParam.ofTool( //
						ChatCompletionToolMessageParam.builder() //
								.content(r.getResult().toString()) //
								.toolCallId(r.getToolCallId()).build() //
				));
			}

			// There is nothing else to return since in this case the message will only have
			// call results.
			return result;
		}

		// In all other cases, we expect a simple text message
		if (msg.isText()) {
			if (msg.getParts().size() != 1)
				throw new IllegalArgumentException(
						"Message can be only be simple text, a tool call, or tool call results");

			// Return corresponding ChatCompletionMessageParam
			switch (msg.getAuthor()) {
			case USER:
				return List.of(ChatCompletionMessageParam.ofUser( //
						ChatCompletionUserMessageParam.builder().content(msg.getTextContent()).build() //
				));
			case DEVELOPER:
				return List.of(ChatCompletionMessageParam.ofDeveloper( //
						ChatCompletionDeveloperMessageParam.builder().content(msg.getTextContent()).build() //
				));
			case BOT:
				return List.of(ChatCompletionMessageParam.ofAssistant( //
						ChatCompletionAssistantMessageParam.builder().content(msg.getTextContent()).build() //
				));
			default:
				throw new IllegalArgumentException("Message author not supported: " + msg.getAuthor());
			}
		}

		throw new IllegalArgumentException("Message can be only be simple text, a tool call, or tool call results");
	}

	/**
	 * @return A ResponseFormat that can be used to create a
	 *         ChatCompletionCreateParams to call OpenAI Chat Completion API.
	 */
	private ChatCompletionCreateParams.ResponseFormat createResponseFormat() {

		try {
			return ResponseFormat.ofJsonSchema( //
					ResponseFormatJsonSchema.builder() //
							.jsonSchema( //
									ResponseFormatJsonSchema.JsonSchema.builder() //
											.name("Anonymous") //
											.description("No Description") //
											.schema(JsonSchema.deserialize(getResponseFormat(), Schema.class)) //
											.build() //
							).build());
		} catch (JsonProcessingException e) {
			// Rethrows as runtime exception; in Python you do not need to know since you do
			// not need to declare exceptions in methods
			throw new RuntimeException(e);
		}
	}

	/**
	 * @return A List<ChatCompletionTool> that can be used to create a
	 *         ChatCompletionCreateParams to call OpenAI Chat Completion API. The
	 *         list will contain definitions for all the tools this agent can use.
	 */
	private List<ChatCompletionTool> createToolDefinitions() {

		List<Tool> tools = new ArrayList<>(toolMap.values());
		List<ChatCompletionTool> tls = new ArrayList<>(tools.size());
		for (Tool t : tools) {
			ChatCompletionTool f;
			try {
				f = ChatCompletionTool.builder() //
						.function( //
								FunctionDefinition.builder() //
										.name(t.getId()) //
										.description(t.getDescription()) //
										.strict(false) //
										.parameters(
												JsonSchema.deserialize(t.getJsonParameters(), FunctionParameters.class))
										.build() //
						).build();
			} catch (JsonProcessingException e) {
				// Rethrows as runtime exception; in Python you do not need to know since you do
				// not need to declare exceptions in methods
				throw new RuntimeException(e);
			}
			tls.add(f);
		}

		return tls;
	}

	/**
	 * Closes the agent freeing up resources.
	 */
	public void close() {
		for (Tool t : toolMap.values()) {
			try {
				t.close();
			} catch (Exception e) {
			}
		}

		try {
			client.close();
		} catch (Exception e) {
		}
	}

	public static class Format {
		@JsonProperty(required = true)
		@JsonPropertyDescription("Your output message")
		public String out;
	}

	public static void main(String[] args) {
		Agent bot = null;
		try {
			bot = new Agent();
			bot.setPersonality(
					"Always output messages as described by this schema: " + JsonSchema.getJsonSchema(Format.class));
			bot.setResponseFormat(Format.class);

			System.out.println(bot.chat("Ciao!").getText());
		} catch (Exception e) {
			e.printStackTrace(System.err);
		} finally {
			if (bot != null)
				try {
					bot.close();
				} catch (Exception e) {
				}
		}
	}
}

File: Nikhil/python/agent.py
"""agent.py

Python translation of the Java `Agent` class.

The implementation mirrors the original behaviour while embracing Pythonic
conventions and the `openai` Python SDK.

Dependencies
------------
- openai                (official OpenAI SDK)
- json                  (standard library)
- logging               (standard library)
- chat_types.ChatMessage, chat_types.ChatCompletion,
  chat_types.ToolCall, chat_types.ToolCallResult, chat_types.TextPart
- tool.Tool             (protocol / abstract base class for tools)
- json_schema.JsonSchema
"""

from __future__ import annotations

import json
import logging
import os
from collections.abc import Iterable, Sequence
from typing import Any, Dict, List, Mapping, MutableMapping, Type

import openai

from chat_types import (
    ChatCompletion,
    ChatMessage,
    TextPart,
    ToolCall,
    ToolCallResult,
)
from json_schema import JsonSchema
from tool import Tool

# --------------------------------------------------------------------------- #
# Logging (equivalent to Java SimpleLogger)
# --------------------------------------------------------------------------- #
logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(name)s [%(levelname)s] %(message)s",
)
logger = logging.getLogger(__name__)

# --------------------------------------------------------------------------- #
# Agent
# --------------------------------------------------------------------------- #
class Agent:
    """
    An *Agent* that uses OpenAI’s Chat Completions API.

    Parameters
    ----------
    id_ : str
        Unique identifier for the agent.
    description : str
        Human-readable description of the agent’s capabilities.
    tools : Iterable[Tool], optional
        Tools that the agent can invoke.
    """

    DEFAULT_MODEL: str = "gpt-4.1"

    # -------------------------- construction --------------------------- #
    def __init__(
        self,
        id_: str = "OpenAIChatCompletionService",
        description: str = "Test agent",
        tools: Iterable[Tool] | None = None,
    ) -> None:
        if id_ is None:
            raise ValueError("id_ must not be None")
        if description is None:
            raise ValueError("description must not be None")

        self.id: str = id_
        self.description: str = description

        # Tools --------------------------------------------------------- #
        self._tool_map: Dict[str, Tool] = {}
        for tool in tools or []:
            if tool is None:
                raise ValueError("tools must not contain None")
            tool.init(self)
            self._tool_map[tool.id] = tool

        # Conversation state ------------------------------------------- #
        self.history: List[ChatMessage] = []
        self.max_history_length: int = float("inf")  # no hard limit
        self.max_conversation_steps: int = float("inf")

        # Model configuration ------------------------------------------ #
        self.model: str = self.DEFAULT_MODEL
        self.temperature: float = 0.0
        self.personality: str | None = None
        self._response_format: str | None = None

        # OpenAI configuration ----------------------------------------- #
        # Expect OPENAI_API_KEY in the environment
        openai.api_key = os.getenv("OPENAI_API_KEY")

    # --------------------------- utils -------------------------------- #
    # Conversation helpers ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
    def clear_conversation(self) -> None:
        """Start a new chat (clears stored history)."""
        self.history.clear()

    # Personality / response-format ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
    @property
    def response_format(self) -> str | None:
        return self._response_format

    def set_response_format(self, schema: Type) -> None:
        """Define an explicit JSON schema for model outputs."""
        if schema is None:
            raise ValueError("schema must not be None")
        self._response_format = JsonSchema.get_json_schema(schema)

    # --------------------- public chat API ---------------------------- #
    def chat(self, message: str | ChatMessage | Sequence[ChatMessage]) -> ChatCompletion:
        """
        Continue the ongoing conversation with *message*.

        The provided message(s) are appended to the conversation, the LLM is
        queried, and the reply is stored in the history.
        """
        # Normalise input
        if isinstance(message, str):
            new_messages = [ChatMessage(message)]
        elif isinstance(message, ChatMessage):
            new_messages = [message]
        else:
            new_messages = list(message)

        # Build conversation context
        conversation: List[ChatMessage] = list(self.history) + new_messages
        self._trim_conversation(conversation)

        # Call the model
        completion = self._chat_completion(conversation)

        # Update history (respecting max_history_length)
        self.history.extend(new_messages)
        self.history.append(completion.message)
        if len(self.history) > self.max_history_length:
            del self.history[: len(self.history) - self.max_history_length]

        return completion

    # ------------------------------------------------------------------ #
    # One-shot completion (ignores history) ---------------------------- #
    def complete(self, prompt: str | ChatMessage) -> ChatCompletion:
        """Run *prompt* outside the conversation (history is untouched)."""
        single = ChatMessage(prompt) if isinstance(prompt, str) else prompt
        conversation = [single]
        self._trim_conversation(conversation)
        return self._chat_completion(conversation)

    # -------------------------- internals ----------------------------- #
    # Trim conversation to honour limits and add personality ~~~~~~~~~~~ #
    def _trim_conversation(self, messages: List[ChatMessage]) -> None:
        """Mutate *messages* so it respects configured limits."""
        # Remove leading tool-results without matching calls
        while messages and messages[0].has_tool_call_results():
            messages.pop(0)

        # Enforce max steps
        if len(messages) > self.max_conversation_steps:
            del messages[: len(messages) - self.max_conversation_steps]

        if not messages:
            raise ValueError("No messages left in conversation after trimming")

        # Inject personality (developer role) as first message
        if self.personality:
            messages.insert(0, ChatMessage(self.personality, ChatMessage.Author.DEVELOPER))

    # Convert ChatMessage → OpenAI message dict ~~~~~~~~~~~~~~~~~~~~~~~~ #
    def _from_chat_message(self, msg: ChatMessage) -> List[Dict[str, Any]]:
        if msg.has_tool_calls():
            tool_calls = []
            for call in msg.get_tool_calls():
                tool_calls.append(
                    {
                        "id": call.id,
                        "type": "function",
                        "function": {
                            "name": call.tool.id,
                            "arguments": json.dumps(call.arguments, separators=(",", ":")),
                        },
                    }
                )
            return [{"role": "assistant", "content": None, "tool_calls": tool_calls}]

        if msg.has_tool_call_results():
            return [
                {
                    "role": "tool",
                    "tool_call_id": r.tool_call_id,
                    "content": r.result,
                }
                for r in msg.get_tool_call_results()
            ]

        # Plain text
        if not msg.is_text():
            raise ValueError("Message must be text, a tool call, or tool call results")

        role_map = {
            ChatMessage.Author.USER: "user",
            ChatMessage.Author.BOT: "assistant",
            ChatMessage.Author.DEVELOPER: "system",  # OpenAI 'developer' maps to 'system'
        }
        return [{"role": role_map[msg.author], "content": msg.get_text_content()}]

    # Convert OpenAI message → ChatMessage ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
    def _from_openai_message(self, message: Mapping[str, Any]) -> ChatMessage:
        if "tool_calls" in message:
            calls: List[ToolCall] = []
            for tc in message["tool_calls"]:
                tool_id = tc["function"]["name"]
                tool = self._tool_map.get(tool_id)
                if tool is None:
                    raise ValueError(f"No tool registered with id '{tool_id}'")

                calls.append(
                    ToolCall(
                        id_=tc["id"],
                        tool=tool,
                        arguments=json.loads(tc["function"]["arguments"]),
                    )
                )
            return ChatMessage(ChatMessage.Author.BOT, calls)

        parts: List[TextPart] = []
        if content := message.get("content"):
            parts.append(TextPart(str(content)))
        if message.get("role") == "assistant" and message.get("content") is None:
            parts.append(TextPart("**The model generated an empty response**"))

        return ChatMessage(ChatMessage.Author.BOT, parts)

    # Prepare response_format / tools for OpenAI call ~~~~~~~~~~~~~~~~~~ #
    def _create_response_format(self) -> Dict[str, Any] | None:
        if self._response_format is None:
            return None
        return {
            "type": "json_object",
            "schema": json.loads(self._response_format),
        }

    def _create_tool_definitions(self) -> List[Dict[str, Any]] | None:
        if not self._tool_map:
            return None
        tools = []
        for t in self._tool_map.values():
            tools.append(
                {
                    "type": "function",
                    "function": {
                        "name": t.id,
                        "description": t.description,
                        "parameters": json.loads(t.json_parameters),
                    },
                }
            )
        return tools

    # Core: call OpenAI and wrap result ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
    def _chat_completion(self, messages: Sequence[ChatMessage]) -> ChatCompletion:
        openai_messages: List[Dict[str, Any]] = []
        for m in messages:
            openai_messages.extend(self._from_chat_message(m))

        req: Dict[str, Any] = {
            "model": self.model,
            "messages": openai_messages,
            "temperature": self.temperature,
        }

        if (rf := self._create_response_format()) is not None:
            req["response_format"] = rf
        if (td := self._create_tool_definitions()) is not None:
            req["tools"] = td

        logger.info("OpenAI request: %s", req)

        resp = openai.ChatCompletion.create(**req)
        choice = resp.choices[0]
        finish_reason = self._map_finish_reason(choice.finish_reason)

        chat_message = self._from_openai_message(choice.message)
        return ChatCompletion(finish_reason, chat_message)

    # Finish-reason mapping ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
    @staticmethod
    def _map_finish_reason(finish: str | None) -> ChatCompletion.FinishReason:
        mapping = {
            "stop": ChatCompletion.FinishReason.COMPLETED,
            "tool_calls": ChatCompletion.FinishReason.COMPLETED,
            "function_call": ChatCompletion.FinishReason.COMPLETED,
            "length": ChatCompletion.FinishReason.TRUNCATED,
            "content_filter": ChatCompletion.FinishReason.INAPPROPRIATE,
        }
        return mapping.get(finish, ChatCompletion.FinishReason.OTHER)

	# Finish-reason mapping ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #      
    @staticmethod
    def fill_slots(template: str, slots: Mapping[str, Any]) -> str:  
        """
        Replace every ``{{key}}`` in *template* with the corresponding value
        from *slots*. Unknown keys are left untouched; *None* is replaced by
        the empty string.

        Parameters
        ----------
        template : str
            A string containing ``{{placeholders}}``.
        slots : Mapping[str, Any]
            Values to inject into the template.

        Returns
        -------
        str
            The rendered string.
        """
        if template is None:
            raise ValueError("template must not be None")
        if slots is None:
            raise ValueError("slots must not be None")

        def _sub(match: re.Match[str]) -> str:
            key = match.group(1).strip()
            value = slots.get(key)
            return "" if value is None else str(value)

        return re.sub(r"\{\{([^}]+)\}\}", _sub, template)
        
    # --------------------------- teardown ----------------------------- #
    def close(self) -> None:
        """Close all tools; nothing required for *openai* client."""
        for tool in self._tool_map.values():
            try:
                tool.close()
            except Exception:  # noqa: BLE001
                logger.exception("Error while closing tool %s", tool.id)
